src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -126,6 +125,9 @@
     private final ZooKeeperServerListener listener;
     private ZooKeeperServerShutdownHandler zkShutdownHandler;
 
+    private volatile long lastSnapshotZxid;
+    private AtomicInteger isGeneratingSnapshot = new AtomicInteger(0);
```

Comment:
In fact, I'm not sure whether Zookeeper backend threads will take snapshot in parallel. So, I choose to use a `AtomicInteger` to protect manual call `takeSnapshot`.